domino/domino.py

```diff
@@ -53,6 +56,16 @@ def __init__(self, project, api_key=None, host=None, domino_token_file=None, aut
             self._logger.error(error_message)
             raise Exception(error_message)
 
+    def _authenticate(fn):
+        """
+        Decorator for automatic re-authentication
+        """
+        def wrapper(self, *args, **kwargs):
```

Comment:
>> The decorator assumes only one type of authentication -- token file. This is the my concern.

Yes and this is by design. There's nothing to do with the API key or auth_token at present time. We can rename `_authenticate` to something else to clarify it is only applicable to the token file case.

Reply:
I ran the unit tests on the `master` branch, and all tests passed.
```
platform linux -- Python 3.6.9, pytest-6.2.5, py-1.10.0, pluggy-1.0.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /mnt/python-domino, configfile: pytest.ini
plugins: requests-mock-1.9.3
collected 11 items                                                                                                                                             

tests/test_basic_auth.py::test_no_auth_type_error PASSED
tests/test_basic_auth.py::test_invalid_token_file_error PASSED
tests/test_basic_auth.py::test_malformed_project_input_error PASSED
tests/test_basic_auth.py::test_object_creation_with_auth_token PASSED
tests/test_basic_auth.py::test_object_creation_with_api_key PASSED
tests/test_basic_auth.py::test_object_creation_with_token_file PASSED
tests/test_basic_auth.py::test_re_authentication_with_a_new_type PASSED
tests/test_basic_auth.py::test_auth_token_precedence PASSED
tests/test_basic_auth.py::test_token_file_precedence PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_api_key PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_token_file PASSED
```
And then again on the `reauthn` branch.
```
tests/test_basic_auth.py::test_no_auth_type_error PASSED
tests/test_basic_auth.py::test_invalid_token_file_error PASSED
tests/test_basic_auth.py::test_malformed_project_input_error PASSED
tests/test_basic_auth.py::test_object_creation_with_auth_token PASSED
tests/test_basic_auth.py::test_object_creation_with_api_key PASSED
tests/test_basic_auth.py::test_object_creation_with_token_file PASSED
tests/test_basic_auth.py::test_re_authentication_with_a_new_type PASSED
tests/test_basic_auth.py::test_auth_token_precedence FAILED   # <----- this now fails
tests/test_basic_auth.py::test_token_file_precedence PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_api_key PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_token_file PASSED
```
...and as suspected, at least one test now fails. (I ran the tests in a Domino workspace. If I installed on my laptop and ran from there, I suspect more than one test would fail.)

I think this is the issue. As the class is defined now, the authentication type is one of the initial setup parameters.
```python
class Domino:
    def __init__(self, project, api_key=None, host=None, domino_token_file=None, auth_token=None):
```
So if I instantiate a new python-domino client like this:
```python
domino = Domino(host="https://my.host.com", project="me/quick-start", api_key="my_secret_key")
```
...the user's expectation (rightly so) is that their choice to authenticate with an API key is part of the saved state of that object.

But the decorator obliterates that initial input parameter by always forcing only one kind of authentication -- token file only.

In summary, I guess my general feeling is this is a good start, but it feels to me like more of a quick fix with "happy path" testing to solve one specific issue, without holistically taking into consideration the whole class.

Reply:
Well okay, if you're saying that precedence is real and not artificial thing (never heard of anyone really supplying more than one artifact for authentication purposes), then of course we should do what I mentioned before.

Reply:
I agree that more supplying more than on authentication method at one time would be weird. If you do so now, the order of precedence is 1) API key, 2) token file, 3) auth token string -- but in general, it should not be done.

However, deciding to use API token (for example) is a real thing. And in fact, all of the tests QE runs to validate python-domino use that method, and they are most often run from a test client that is not part of the deployment.

After your most recent commit, unit tests are now passing if run from within a workspace. My feeling is still that this is a large change (because it inserts a new method call each and every time the API is accessed), so I personally still need some time to consider the full implications.